emu/cores/pokey.c

```diff
@@ -165,1028 +330,727 @@ const DEV_DEF* devDefList_Pokey[] =
-#define IRQ_TIMR1	0x01	/* timer channel #1 interrupt */
+	INIT_DEVINF(retDevInf, &d->_devData, cfg->clock, &devDef);
+	return 0x00;
+}
 
-/* SKSTAT (R/D20F) */
-#define SK_FRAME	0x80	/* serial framing error */
-#define SK_OVERRUN	0x40	/* serial overrun error */
-#define SK_KBERR	0x20	/* keyboard overrun error */
-#define SK_SERIN	0x10	/* serial input high */
-#define SK_SHIFT	0x08	/* shift key pressed */
-#define SK_KEYBD	0x04	/* keyboard key pressed */
-#define SK_SEROUT	0x02	/* serial output active */
+static void pokey_device_reset(void *info)
+{
+	pokey_device *d = (pokey_device *)info;
 
-/* SKCTL (W/D20F) */
-#define SK_BREAK	0x80	/* serial out break signal */
-#define SK_BPS		0x70	/* bits per second */
-#define SK_FM		0x08	/* FM mode */
-#define SK_PADDLE	0x04	/* fast paddle a/d conversion */
-#define SK_RESET	0x03	/* reset serial/keyboard interface */
+	/* Setup channels */
+	for (int i=0; i<POKEY_CHANNELS; i++)
+	{
+		d->m_channel[i].m_parent = d;
+		d->m_channel[i].m_INTMask = 0;
+		d->m_channel[i].m_AUDF = 0;
+		d->m_channel[i].m_AUDC = 0;
+		d->m_channel[i].m_borrow_cnt = 0;
+		d->m_channel[i].m_counter = 0;
+		d->m_channel[i].m_output = 0;
+		d->m_channel[i].m_filter_sample = 0;
+	}
+	d->m_channel[CHAN1].m_INTMask = IRQ_TIMR1;
+	d->m_channel[CHAN2].m_INTMask = IRQ_TIMR2;
+	d->m_channel[CHAN4].m_INTMask = IRQ_TIMR4;
+
+	// bind callbacks
+	/*
+	m_keyboard_r.resolve();
+	m_irq_f.resolve();
+	*/
+
+	/* calculate the A/D times
+	 * In normal, slow mode (SKCTL bit SK_PADDLE is clear) the conversion
+	 * takes N scanlines, where N is the paddle value. A single scanline
+	 * takes approximately 64us to finish (1.78979MHz clock).
+	 * In quick mode (SK_PADDLE set) the conversion is done very fast
+	 * (takes two scanlines) but the result is not as accurate.
+	 */
 
-#define DIV_64		28		 /* divisor for 1.78979 MHz clock to 63.9211 kHz */
-#define DIV_15		114 	 /* divisor for 1.78979 MHz clock to 15.6999 kHz */
+	/* initialize the poly counters */
+	poly_init_4_5(d->m_poly4, 4, 1, 0);
+	poly_init_4_5(d->m_poly5, 5, 2, 1);
+
+	/* initialize 9 / 17 arrays */
+	poly_init_9_17(d->m_poly9,   9);
+	poly_init_9_17(d->m_poly17, 17);
+	pokey_device_vol_init(d);
+
+	/* The pokey does not have a reset line. These should be initialized
+	 * with random values.
+	 */
+
+	d->m_KBCODE = 0x09;		 /* Atari 800 'no key' */
+	d->m_SKCTL = SK_RESET;  /* let the RNG run after reset */
+	d->m_SKSTAT = 0;
+	/* This bit should probably get set later. Acid5200 pokey_setoc test tests this. */
+	d->m_IRQST = IRQ_SEROC;
+	d->m_IRQEN = 0;
+	d->m_AUDCTL = 0;
+	d->m_p4 = 0;
+	d->m_p5 = 0;
+	d->m_p9 = 0;
+	d->m_p17 = 0;
+	d->m_ALLPOT = 0x00;
+
+	d->m_pot_counter = 0;
+	d->m_kbd_cnt = 0;
+	d->m_out_filter = 0;
+	d->m_out_raw = 0;
+	d->m_old_raw_inval = 1;
+	d->m_kbd_state = 0;
+
+	d->m_icount = 0;
+
+	/* reset more internal state */
+	memset(d->m_clock_cnt,0,sizeof(d->m_clock_cnt));
+	memset(d->m_clock_cnt,0,sizeof(d->m_POTx));
```

Comment:
I think this should clear `d->m_POTx`.

Reply:
Oops! Thanks for that catch, fixed in 3ceaf98b3bbbfcd48c8107cb903aa40f267c3e55